### ISSUE TITLE
How to Add an LSTM Layer for Sequential Pose Detection to YOLOv11x-Pose Model with Custom Keypoints Data

### ISSUE THREAD

**Issue by Rasantis** [ultralytics/ultralytics#16673]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi Ultralytics Team,

I’m currently working on a complex project where I’m using the yolo11x-pose.pt model to detect human poses and keypoints, and I’m trying to enhance it by adding an LSTM layer to perform sequential analysis on these pose keypoints. The goal is to identify suspicious actions like theft over time by feeding the detected keypoints into an LSTM for temporal pattern recognition. Here’s a breakdown of what we’re trying to achieve and where we are facing difficulties.

Context and Current Workflow:
Pose Detection Using YOLOv11x-Pose:

We are using yolo11x-pose.pt to detect human poses (keypoints) from videos.
The current model is successfully extracting keypoints from the videos. Each frame outputs coordinates (x, y) and confidence for keypoints like head, shoulders, elbows, knees, etc.

Keypoints Data Format:

The data we are extracting from each video frame is formatted as follows (example from one frame):
[  [[x1, y1, confidence1], [x2, y2, confidence2], ...],
  ...
]
Each frame provides 17 keypoints per person detected in the scene, and we are tracking multiple people over time.

Objective:

We want to pass these keypoints (ignoring confidence values) over a sequence of frames (e.g., 120 frames) into an LSTM layer.
The LSTM layer should analyze the sequence to detect suspicious behavior (like theft) based on the temporal pattern of movement.

Current Approach:
Keypoints Extraction:

We are extracting the keypoints successfully from the YOLOv11x-pose model over multiple frames.
LSTM Layer for Temporal Analysis:

We’ve designed a basic LSTM model that takes sequences of keypoints as input.
The idea is to use the LSTM to detect suspicious actions based on how a person’s keypoints move over time. We want to predict a class (e.g., "theft" or "normal") based on these sequences.

Challenges and Questions:
Now, we want to integrate this sequential LSTM layer with the yolo11x-pose model directly so that:

The LSTM Layer Receives Sequential Keypoints:

Instead of processing frames independently, we want the LSTM to process the sequence of frames for each detected person.
For each detected person, track their keypoints over a sequence (e.g., 120 frames) and feed that into the LSTM.
Modifying YOLOv11x-Pose:

We are unsure how to modify the yolo11x-pose.pt model to allow for temporal tracking and LSTM integration.
Specifically, how can we append the LSTM layer to the current architecture so that it takes the keypoints for each person and processes them as a temporal sequence?
Do we need to extract keypoints outside the model and feed them into a separate LSTM, or can we add the LSTM directly into the YOLO pipeline?
Training Workflow:

How should we set up the training pipeline for this combined architecture?
Should we perform the LSTM training separately after keypoint extraction, or is there a way to jointly train both the pose model and the LSTM layer?
Fine-tuning:

How do we fine-tune the combined YOLOv11x-pose model with the LSTM layer on our custom dataset of theft actions?
The goal is to train the LSTM part with sequences of keypoints from theft actions, but we also need to ensure the keypoint detection remains accurate.

Goal:
Our end goal is to combine the pose detection from YOLOv11x-pose with an LSTM layer for sequential action recognition to detect theft patterns. We believe this combination could provide high accuracy in real-world scenarios, but we are struggling with integrating the components smoothly.

Any guidance on how to properly extend the YOLOv11x-pose model to include the LSTM layer for sequence processing, or how to structure the workflow, would be greatly appreciated!

Thanks in advance for your help.

### Additional

I´m going to send some of my scripts:
[keypoints_12_15_2024_10_03_02_06_33.json](https://github.com/user-attachments/files/17248899/keypoints_12_15_2024_10_03_02_06_33.json)
[keypoints_12_15_2024_10_03_02_06_43.json](https://github.com/user-attachments/files/17248900/keypoints_12_15_2024_10_03_02_06_43.json)
[keypoints_8_10_2024_10_03_02_10_46.json](https://github.com/user-attachments/files/17248901/keypoints_8_10_2024_10_03_02_10_46.json)
my python scripts:
data_preparation.py:
import cv2
import json
import torch
from ultralytics import YOLO
import time

# Definindo as conexões entre keypoints focadas nos ombros, braços e quadris
SKELETON_CONNECTIONS = [
    (5, 7), (7, 9),  # Ombro esquerdo para cotovelo esquerdo e pulso esquerdo
    (6, 8), (8, 10),  # Ombro direito para cotovelo direito e pulso direito
    (5, 11), (6, 12),  # Ombros para quadris
]

# Função para desenhar keypoints e conectar os pontos (foco nos membros superiores e quadris)
def draw_keypoints_and_skeleton(frame, keypoints, conf_threshold=0.5):
    """
    Desenha os keypoints focados em braços e quadris, conectando-os na imagem.
    
    :param frame: Frame de vídeo onde desenhar os keypoints.
    :param keypoints: Lista de keypoints extraídos.
    :param conf_threshold: Limite de confiança para desenhar keypoints (padrão 0.5).
    :return: Frame com keypoints e esqueleto desenhados.
    """
    valid_keypoints = []
    
    # Extrai e processa apenas os keypoints relevantes (membros superiores e quadris)
    for i, keypoint in enumerate(keypoints[0]):  # Considerando que keypoints é uma lista de listas
        x, y, conf = keypoint  # Extraímos as coordenadas e a confiança
        if x != 0 and y != 0 and conf >= conf_threshold:  # Verifica se o keypoint é válido
            valid_keypoints.append((int(x), int(y)))
            cv2.circle(frame, (int(x), int(y)), 5, (0, 255, 0), -1)  # Desenha um círculo verde
        else:
            valid_keypoints.append(None)  # Marca como inválido se não atender aos critérios

    # Desenhar as linhas conectando os keypoints focados nos membros superiores e quadris
    for connection in SKELETON_CONNECTIONS:
        kp1, kp2 = connection
        if valid_keypoints[kp1] and valid_keypoints[kp2]:  # Se ambos os keypoints são válidos
            cv2.line(frame, valid_keypoints[kp1], valid_keypoints[kp2], (255, 0, 0), 2)  # Desenha a linha azul

    return frame

# Função para salvar keypoints em um arquivo JSON
def save_keypoints(keypoints, output_path):
    """
    Salva os keypoints extraídos em um arquivo JSON.
    
    :param keypoints: Lista de keypoints extraídos.
    :param output_path: Caminho para salvar o arquivo JSON.
    """
    with open(output_path, 'w') as f:
        json.dump(keypoints.tolist(), f)  # Converte tensor para lista e salva
    print(f"Keypoints salvos em {output_path}")

# Função principal para extrair keypoints de um vídeo
def extract_keypoints(video_path, output_folder, start_time=0, end_time=None, conf_threshold=0.5):
    """
    Extrai keypoints de um vídeo entre um intervalo de tempo especificado.
    
    :param video_path: Caminho para o vídeo.
    :param output_folder: Pasta para salvar os keypoints extraídos.
    :param start_time: Tempo inicial para começar a análise (em segundos).
    :param end_time: Tempo final para terminar a análise (em segundos). Se None, processa até o fim do vídeo.
    :param conf_threshold: Limite de confiança para desenhar keypoints (padrão 0.5).
    """
    model = YOLO('modelos/yolo11x-pose.pt')  # Carrega o modelo YOLOv8 Pose
    #SHOPPLIFTING VIDEO
    # Carrega o vídeo
    cap = cv2.VideoCapture(video_path)
    
    fps = cap.get(cv2.CAP_PROP_FPS)
    total_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
    
    start_frame = int(start_time * fps)
    end_frame = int(end_time * fps) if end_time else total_frames
    
    # Verifica se o vídeo foi carregado corretamente
    if not cap.isOpened():
        print("Erro ao abrir o vídeo!")
        return
    
    frame_idx = 0
    keypoints_data = []

    while cap.isOpened():
        ret, frame = cap.read()
        if not ret or frame_idx > end_frame:
            break

        if frame_idx >= start_frame:
            results = model(frame, max_det=1)  # Realiza a inferência

            if len(results) > 0:
                keypoints = results[0].keypoints.data.cpu()  # Extrai os keypoints e move para a CPU
                keypoints_data.append(keypoints.tolist())  # Armazena keypoints para salvar depois

                print(f"Keypoints extraídos do frame {frame_idx}: {keypoints.tolist()}")
                frame_with_skeleton = draw_keypoints_and_skeleton(frame, keypoints, conf_threshold)  # Desenha esqueleto

                # Exibe a imagem com os keypoints e esqueleto desenhados (opcional)
                cv2.imshow('Keypoints and Skeleton', frame_with_skeleton)
                if cv2.waitKey(1) & 0xFF == ord('q'):
                    break
        
        frame_idx += 1

    # Obter o timestamp atual no formato ano_mes_dia_hora_min_seg
    timestamp = time.strftime("%Y_%m_%d_%H_%M_%S")

    # Salva os keypoints em um arquivo JSON com o timestamp
    save_keypoints(torch.tensor(keypoints_data), f"{output_folder}/keypoints_{start_time}_{end_time}_{timestamp}.json")
        
    # Libera os recursos
    cap.release()
    cv2.destroyAllWindows()

# Executa a função de extração com os parâmetros fornecidos
if __name__ == "__main__":
    video_path = 'vds_treinamento/video_furto4.mp4'  # Substitua pelo caminho do seu vídeo
    output_folder = 'keypoints_data'  # Pasta onde os keypoints serão salvos
    extract_keypoints(video_path, output_folder, start_time=41, end_time=45)  # Ajuste o intervalo conforme necessário

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Rasantis, thank you for your interest in Ultralytics 🚀! Your project sounds fascinating and ambitious! Here's some information and resources to help you get started.

For complex modifications, we recommend consulting the [Docs](https://docs.ultralytics.com), especially the sections on [Custom Training](https://docs.ultralytics.com/guides/model-training-tips/) and [Python Usage](https://docs.ultralytics.com/usage/python/), which may provide insights into handling model modifications, including adding custom layers.

As you're working on adding an LSTM layer to the YOLOv11x-pose model for temporal analysis, it can be quite intricate. Since this extends beyond typical use, consider providing a 🐛 [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) if you face specific bug-related challenges.

For questions on extending the model's architecture, please include detailed descriptions and scripts, as you've done, to aid in troubleshooting. Keep experimenting, as real-time analysis can be complex.

Meanwhile, join our community for more dynamic discussions and support:
- [Discord](https://ultralytics.com/discord) for real-time chat 🎧
- [Discourse](https://community.ultralytics.com) for in-depth discussions
- [Subreddit](https://reddit.com/r/ultralytics) to share and gain insights

## Upgrade

Ensure you are using the latest package and dependencies. Upgrade using:

```bash
pip install -U ultralytics
```

Additionally, use the recommended [environments](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) for smooth operation, especially with custom modifications.

## Status

If you need assurance of our continuous integration, check:

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

This is an automated response, and an Ultralytics engineer will review your query soon. Keep experimenting, and we'll assist you in combining YOLO with sequential analysis for your intriguing use-case! 😊

**Comment by Skillnoob** [comment]
You'd need to add support for that layer in one of the files of [https://github.com/ultralytics/ultralytics/tree/main/ultralytics/nn/modules](this folder) and then modify the yolo11-pose yaml to include this layer.

**Comment by glenn-jocher** [comment]
Hi @Hemilibeatriz, 

For integrating LSTM with YOLO11-pose, I'd recommend a two-stage approach rather than modifying the core YOLO architecture. Your current method of extracting keypoints is already on the right track. Here's how to proceed:

1. Keep using YOLO11-pose to extract keypoints frame-by-frame as you're doing
2. Process these keypoints with person tracking to maintain identity across frames
3. Create a separate LSTM network that takes sequences of keypoints as input

This approach is more maintainable than modifying YOLO's internal architecture. You can implement this by:

```python
# After extracting keypoints with YOLO11
sequences = process_and_track_keypoints(keypoints_data)  # Group by person ID

# Define LSTM model
lstm_model = nn.Sequential(
    nn.LSTM(input_size=34, hidden_size=128, num_layers=2, batch_first=True),
    nn.Linear(128, num_action_classes)
)

# Train LSTM separately on sequences
for sequence in sequences:
    # Each sequence is keypoints for one person over multiple frames
    predictions = lstm_model(sequence)
```

This modular design gives you more flexibility for training and fine-tuning each component separately. If you need tracking implementation details, let me know.
